balancer-js/src/modules/zaps/bbausd2-migrations/stables.ts

```diff
@@ -0,0 +1,250 @@
+   * @param {boolean}                   staked Indicates whether tokens are initially staked or not.
+   * @param {string[]}                  underlyingTokens Underlying token addresses. Array must have the same length and order as underlying tokens in pool being migrated from. Refer to [getPoolTokens](https://github.com/balancer-labs/balancer-v2-monorepo/blob/master/pkg/interfaces/contracts/vault/IVault.sol#L334).
+   * @param {string}                    authorisation Encoded authorisation call.
+   * @returns Migration transaction request ready to send with signer.sendTransaction
+   */
+  calldata(
+    userAddress: string,
+    from: { id: string; address: string; gauge?: string },
+    to: { id: string; address: string; gauge?: string },
+    bptIn: string,
+    minBptOut: string,
+    staked: boolean,
+    underlyingTokens: string[],
+    authorisation?: string
+  ): {
+    to: string;
+    data: string;
+  } {
+    const relayer = this.addresses.relayer;
+    let calls: string[] = authorisation
+      ? [this.buildSetRelayerApproval(authorisation)]
+      : [];
+
+    if (staked && (from.gauge == undefined || to.gauge == undefined))
+      throw new Error(
+        'Staked flow migration requires gauge addresses to be provided'
+      );
+
+    if (staked) {
+      calls = [
+        ...calls,
+        this.buildWithdraw(userAddress, bptIn, from.gauge as string),
+        this.buildExit(from.id, relayer, bptIn, underlyingTokens),
+        this.buildSwap(minBptOut, relayer, to.id, to.address, underlyingTokens),
+        this.buildDeposit(userAddress, to.gauge as string),
+      ];
+    } else {
+      calls = [
+        ...calls,
+        this.buildExit(from.id, userAddress, bptIn, underlyingTokens),
+        this.buildSwap(
+          minBptOut,
+          userAddress,
+          to.id,
+          to.address,
+          underlyingTokens
+        ),
+      ];
+    }
+
+    const callData = balancerRelayerInterface.encodeFunctionData('multicall', [
+      calls,
+    ]);
+
+    return {
+      to: this.addresses.relayer,
+      data: callData,
+    };
+  }
+
+  /**
+   * Encodes exitPool call data.
+   * Exit stable pool proportionally to underlying stables. Exits to relayer.
+   * Outputreferences are used to store exit amounts for next transaction.
+   *
+   * @param {string}    poolId Pool id.
+   * @param {string}    sender Sender address.
+   * @param {string}    amount Amount of BPT to exit with.
+   * @param {string[]}  underlyingTokens Token addresses to exit to.
+   * @returns Encoded exitPool call. Output references.
+   */
+  buildExit(
+    poolId: string,
+    sender: string,
+    amount: string,
+    underlyingTokens: string[]
+  ): string {
+    // Assume gaugeWithdraw returns same amount value
+    const userData = StablePoolEncoder.exitExactBPTInForTokensOut(amount);
+
+    // Ask to store exit outputs for batchSwap of exit is used as input to swaps
+    // TODO: check how does tokens order matter between exits and swaps
```

Comment:
is it checked? i think here it needs to be the exact token order, while in swaps we can swap into BPTs in any order we want.

Reply:
Yeah, this comment is old and should be removed.

Reply:
✅